cpp/examples/parquet/low-level-api/encryption-reader-writer-all-crypto-options.cc

```diff
@@ -0,0 +1,639 @@
+
+  // This vector will hold various encryption configuraions.
+  std::vector<std::shared_ptr<parquet::FileEncryptionProperties>>
+      vector_of_encryption_configurations;
+
+  // Encryption configuration 1: Encrypt all columns and the footer with the same key.
+  // (uniform encryption)
+  parquet::FileEncryptionProperties::Builder file_encryption_builder_1(
+      kFooterEncryptionKey);
+  // Add to list of encryption configurations.
+  vector_of_encryption_configurations.push_back(
+      file_encryption_builder_1.footer_key_metadata("kf")->build());
+
+  // Encryption configuration 2: Encrypt two columns and the footer, with different keys.
+  std::map<std::shared_ptr<parquet::schema::ColumnPath>,
+           std::shared_ptr<parquet::ColumnEncryptionProperties>,
+           parquet::schema::ColumnPath::CmpColumnPath>
+      encryption_cols2;
+  std::shared_ptr<parquet::schema::ColumnPath> path_ptr =
+      parquet::schema::ColumnPath::FromDotString("double_field");
+  std::shared_ptr<parquet::schema::ColumnPath> path_ptr1 =
+      parquet::schema::ColumnPath::FromDotString("float_field");
+  parquet::ColumnEncryptionProperties::Builder encryption_col_builder_20(path_ptr);
+  parquet::ColumnEncryptionProperties::Builder encryption_col_builder_21(path_ptr1);
+  encryption_col_builder_20.key(kColumnEncryptionKey1)->key_id("kc1");
+  encryption_col_builder_21.key(kColumnEncryptionKey2)->key_id("kc2");
+
+  encryption_cols2[path_ptr] = encryption_col_builder_20.build();
+  encryption_cols2[path_ptr1] = encryption_col_builder_21.build();
+
+  parquet::FileEncryptionProperties::Builder file_encryption_builder_2(
+      kFooterEncryptionKey);
+
+  vector_of_encryption_configurations.push_back(
+      file_encryption_builder_2.footer_key_metadata("kf")
+          ->column_properties(encryption_cols2)
+          ->build());
+
+  // Encryption configuration 3: Encrypt two columns, with different keys.
+  // Don’t encrypt footer.
+  // (plaintext footer mode, readable by legacy readers)
+  std::map<std::shared_ptr<parquet::schema::ColumnPath>,
+           std::shared_ptr<parquet::ColumnEncryptionProperties>,
+           parquet::schema::ColumnPath::CmpColumnPath>
+      encryption_cols3;
+  parquet::ColumnEncryptionProperties::Builder encryption_col_builder_30(path_ptr);
+  parquet::ColumnEncryptionProperties::Builder encryption_col_builder_31(path_ptr1);
+  encryption_col_builder_30.key(kColumnEncryptionKey1)->key_id("kc1");
+  encryption_col_builder_31.key(kColumnEncryptionKey2)->key_id("kc2");
+
+  encryption_cols3[path_ptr] = encryption_col_builder_30.build();
+  encryption_cols3[path_ptr1] = encryption_col_builder_31.build();
+  parquet::FileEncryptionProperties::Builder file_encryption_builder_3(
+      kFooterEncryptionKey);
+
+  vector_of_encryption_configurations.push_back(
+      file_encryption_builder_3.footer_key_metadata("kf")
+          ->column_properties(encryption_cols3)
+          ->set_plaintext_footer()
```

Comment:
Why not have another `parquet::FileEncryptionProperties::Builder` constructor that does not take a `footer_key` and infer `plaintext_footer` from the absence of the footer key?
From a user API point of view, the current requirement of needing a footer-key even if the user wants plaintext footer seems odd.

Reply:
Per the spec, footer key is always required - in plaintext footer mode, it is used to sign / tamper-proof the footer. We'll add an API comment to make this clear.